### PR TITLE
WIP: command line rebasing instructions

### PR DESCRIPTION
**Pull request by sbanerjee23**
Related to #9 

@sbillinge I wrote some very basic steps for myself the other day while @chiahaoliu was showing me what to do, so I've added those steps to the README here. They work and it's not hard because I've followed them a couple times now, but I also don't know any of the details, so it would be good if you and Tim check the wording/steps and fix as necessary.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -246,7 +246,51 @@ makes it very easy to compare edits.
...
+   git fetch upstream master
+   ```
+1. Rebase in interactive mode
+
+   ```
+   git rebase upstream/master -i
+   ```
+1. If you have messy commits, flag them with the squash thing. If there's only one
+conflicing commit, just `ctrl-x` out of the interactive window/message. You 
+should see an error message saying that the patch couldn't be applied.
```
Comment (chiahaoliu) [inline README.md]: Maybe more context like?

1. You will be brought into your system editor (depends on your configuration, could be `vim`, `nano` or other ones), with instruction from `git`
```
pick <commit message 1>
pick <commit message 2>
...

# Rebase <your current commit> onto <last commit on upstream/master> (1 command)
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
# d, drop = remove commit
```
As shown in the instruction, you may `squash` a few commits together if you want to keep the history neat. Other options may be used and picking all commits is totally fine too.

Once you are satisfied with the commit history, simply quit your editor (depends on the editor, it is `ctrl+x` in `nano` and `:wq` in `vim`)  and `git` will start rebasing.

If you see `Successfully rebase to...`, congrats, you don't need to do more edits. Just jump to the last step of this section. If you see something like `the patch couldn't be applied..`, then we have to follow the steps below to fix them.

File: README.md [hunk 2]
```diff
@@ -246,7 +246,51 @@ makes it very easy to compare edits.
...
+
+1. Check where the conflict happened
+
+   ```
+   git status
+   ```
+1. Open the conflicting file with the diff >>>>>>>/======= things.
+
```
Comment (chiahaoliu) [inline README.md]: maybe like?

Open the conflicting file with your favorite editor and search for `>>>>>>>` or `=======`

**Comment by sbanerjee23** [comment]
@chiahaoliu I incorporated the suggestions and made some other edits. I suggest that after @sbillinge has a chance to check this first draft, we merge this PR. There is not much point in you making these suggestions to me, because the README was from me writing down steps while you told me what to do, so you're the expert here!

File: README.md [hunk 3]
```diff
@@ -260,31 +260,57 @@ makes it very easy to compare edits.
    ```
    git rebase upstream/master -i
    ```
-1. If you have messy commits, flag them with the squash thing. If there's only one
-conflicing commit, just `ctrl-x` out of the interactive window/message. You 
-should see an error message saying that the patch couldn't be applied.
+
+1. You will be brought into your system editor (depends on your configuration, could be `vim`, 
+   `nano` or other ones), with instruction from `git`.
```
Comment (sbillinge) [inline README.md]: I find these instructions pretty hard to follow.  Step 1, ok, so now I am in the editor, but after that, what am I seeing?  What am I typing?  What is the workflow?  I don't really have any idea with just 
```
pick ....
```

File: README.md [hunk 4]
```diff
@@ -260,31 +260,57 @@ makes it very easy to compare edits.
...
+   # p, pick = use commit
+   # r, reword = use commit, but edit the commit message
+   # e, edit = use commit, but stop for amending
+   # s, squash = use commit, but meld into previous commit
+   # f, fixup = like "squash", but discard this commit's log message
+   # x, exec = run command (the rest of the line) using shell
+   # d, drop = remove commit
+   ```
+
+   As shown in the instructions, you may `squash` a few commits together if you want to keep
```
Comment (sbillinge) [inline README.md]: what instructions?  Aren't these instructions?

File: README.md [hunk 5]
```diff
@@ -260,31 +260,57 @@ makes it very easy to compare edits.
...
-should see an error message saying that the patch couldn't be applied.
+
+1. You will be brought into your system editor (depends on your configuration, could be `vim`, 
+   `nano` or other ones), with instruction from `git`.
+   ```
+   pick <commit message 1>
+   pick <commit message 2>
+   ...
+
+   # Rebase <your current commit> onto <last commit on upstream/master> (1 command)
```
Comment (sbillinge) [inline README.md]: What is all this about?

**Comment by sbanerjee23** [comment]
@sbillinge I don't know enough to improve on these instructions. I follow them and they work. Maybe some terminal screenshots can clarify which code blocks are prompts, but @chiahaoliu is the right person for this, he did a great job showing me what to do in person so I trust he can capture that in a readme. It might not be the worst solution if we don't include command line instructions and we coerce everyone into using PyCharm, which is what you want to happen. If people don't use PyCharm, they can figure it out for themselves, and if they do it incorrectly then their PR won't be merged. Also did you follow these steps from a terminal? I don't know how the steps could possibly make sense if you're not actually seeing the git prompts in a shell. As far as I can remember, what we did that one time in group meeting is identical to this workflow.

**Comment by sbillinge** [comment]
From the sound of it, it we should have Tim be handling this.  We can close this PR and open one on his?

I do think the workflow in PyCharm is much better.   At least for younger members of the group I think I would strongly encourage them all to use it.

Here are a few general words that maybe clarify my concerns with the instructions.   What is not clear is what I am seeing on the screen vs. what I am supposed to be typing etc., so there is no context.  I get that an editor opens when I start interactive, but then it says ``pick <commit mesage 1>``...do I see that or do I type that? If I am in an editor, aren't I editing a file? etc..  So the instructions may just need a few words, like "you will see:" and so on, but the context is kind of important.  Also, I am a bit concerned that if Tim has taught this to Soham but he doesn't know what is going on (other than blindly following instructions that don't make sense) then it is maybe they ar enot the best instructions.  We don't want to rewrite docs that are on the web, but we should link to good docs on the web that explain things in that case, and capture enough of the context that someone can follow without Tim being there.

Thanks for the efforts.  Rebasing is a big part of this workflow (less so when fewer people are working on a project) so we probably have to bite the bullet and learn it properly.
